atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java

```diff
@@ -105,59 +134,83 @@ public boolean isInitialized() {
+                rows,
+                BatchColumnRangeSelection.create(null, null, 1),
+                Long.MAX_VALUE);
+
+        return rows.stream()
+                .map(itMap::get)
+                .filter(Iterator::hasNext)
+                .map(Iterator::next)
+                .map(Map.Entry::getValue)
+                .map(KeyValueServicePuncherStore::getAtlasTimestamp)
+                .findFirst();
+    }
+
+    private static long getAtlasTimestamp(Value value) {
+        return EncodingUtils.decodeUnsignedVarLong(value.getContents());
     }
 
+    // Relies on a full table scan, but should only be used in a cli.
     public static long getMillisForTimestamp(KeyValueService kvs, long timestamp) {
         long timestampExclusive = timestamp + 1;
-        // punch table is keyed by the real value we're trying to find so we have to do a whole table
-        // scan, which is fine because this table should be really small
-        byte[] startRow = EncodingUtils.encodeUnsignedVarLong(Long.MAX_VALUE);
-        EncodingUtils.flipAllBitsInPlace(startRow);
-        RangeRequest rangeRequest =
-                RangeRequest.builder().startRowInclusive(startRow).batchHint(1).build();
-        ClosableIterator<RowResult<Value>> result =
-                kvs.getRange(AtlasDbConstants.PUNCH_TABLE, rangeRequest, timestampExclusive);
-
-        try {
-            if (result.hasNext()) {
-                byte[] encodedMillis = result.next().getRowName();
-                EncodingUtils.flipAllBitsInPlace(encodedMillis);
-                return EncodingUtils.decodeUnsignedVarLong(encodedMillis);
-            } else {
-                return 0L;
-            }
-        } finally {
-            result.close();
+        RangeRequest rangeRequest = RangeRequest.builder().batchHint(1).build();
+        try (ClosableIterator<RowResult<Value>> result =
+                kvs.getRange(AtlasDbConstants.PUNCH_TABLE, rangeRequest, timestampExclusive)) {
+            // With Guava 21.0, can do Streams.stream(result), Streams.stream(row.getCells())
+            return StreamSupport.stream(Spliterators.spliteratorUnknownSize(result, 0), false)
+                    .flatMap(row -> StreamSupport.stream(row.getCells().spliterator(), false))
+                    .map(Map.Entry::getKey)
+                    .mapToLong(KeyValueServicePuncherStore::timestamp)
+                    .max()
+                    .orElse(0L);
         }
     }
 
+    private static long timestamp(Cell cell) {
+        return MILLIS_IN_TWO_WEEKS * EncodingUtils.decodeLittleEndian(cell.getRowName(), 0)
+                + EncodingUtils.decodeUnsignedVarLong(EncodingUtils.flipAllBits(cell.getColumnName()));
+    }
+
+    private static byte[] row(long timeMillis) {
+        return EncodingUtils.encodeLittleEndian(Math.floorDiv(timeMillis, MILLIS_IN_TWO_WEEKS));
+    }
+
+    // floorDiv and floorMod are to ensure that this works properly if the time is somehow negative.
```

Comment:
nit: this should go before `row()` probably

Nice though, I probably wouldn't have thought of this!

Reply:
done